src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java

```diff
@@ -493,8 +579,57 @@ private static List<List<ObjectIdentifier>> getContainedObjectBatches(S3AsyncCli
      */
     @Override
     public void copy(Path source, Path target, CopyOption... options) throws IOException {
+        //
+        // TODO: source and target can belong to any file system (confirmed, see
+        //       https://github.com/awslabs/aws-java-nio-spi-for-s3/issues/135),
+        //       we can not assume they points to S3 objects
+        //
         try {
-            this.copy(null, source, target, options);
+            // If both paths point to the same object, this is a no-op
+            if (source.equals(target)) {
+                return;
+            }
+
+            List<CopyOption> copyOptions = Arrays.asList(options);
+
+            S3Path s3SourcePath = (S3Path) source;
+            S3Path s3TargetPath = (S3Path) target;
+
+            final S3FileSystem fs = ((S3Path)source).getFileSystem();
+            final S3AsyncClient s3Client = fs.client();
+            final String bucketName = fs.bucketName();
+
+            String prefix = s3SourcePath.toRealPath(NOFOLLOW_LINKS).getKey();
+
+            long timeOut = TIMEOUT_TIME_LENGTH_1;
+            final TimeUnit unit = MINUTES;
+            try {
+                List<List<ObjectIdentifier>> keys = getContainedObjectBatches(s3Client, bucketName, prefix, timeOut, unit);
+
+                for (List<ObjectIdentifier> keyList : keys) {
+                    for (ObjectIdentifier objectIdentifier : keyList) {
+                        S3Path resolvedS3TargetPath = s3TargetPath.resolve(objectIdentifier.key().replaceFirst(prefix + S3Path.PATH_SEPARATOR, ""));
+
+                        if (!copyOptions.contains(StandardCopyOption.REPLACE_EXISTING) && exists(s3Client, resolvedS3TargetPath)) {
+                            throw new FileAlreadyExistsException("File already exists at the target key");
+                        }
+
+                        try (S3TransferManager s3TransferManager = S3TransferManager.builder().s3Client(s3Client).build()) {
+                            s3TransferManager.copy(CopyRequest.builder()
+                                    .copyObjectRequest(CopyObjectRequest.builder()
+                                            .checksumAlgorithm(ChecksumAlgorithm.SHA256)
+                                            .sourceBucket(bucketName)
+                                            .sourceKey(objectIdentifier.key())
+                                            .destinationBucket(resolvedS3TargetPath.bucketName())
+                                            .destinationKey(resolvedS3TargetPath.getKey())
+                                            .build())
+                                    .build()).completionFuture().join();
```

Comment:
It is usually better to use `get(timeOut, unit)` otherwise there is a risk that the method will hang indefinitely and never return. The `get` with time out has the advantage that the method will not wait forever but instead throw an exception that the user can catch and recover from if they want to.